Quadriga/src/main/java/edu/asu/spring/quadriga/web/network/NetworkListController.java

```diff
@@ -78,24 +95,66 @@ public String listNetworks(ModelMap model, Principal principal) throws QuadrigaS
-	public String visualizeNetworks(@PathVariable("networkId") String networkId, ModelMap model, Principal principal) throws QuadrigaStorageException, JAXBException {
+	public String visualizeNetworks(@PathVariable("networkId") String networkId, ModelMap model, Principal principal) throws QuadrigaStorageException, JAXBException {	
+	    // Identify the User
+	    String userId = principal.getName();
+	
+	    // Get network details using the networkId
 		INetwork network = networkManager.getNetwork(networkId);
-		if(network==null){
-			return "auth/404";
-		}
-		ITransformedNetwork transformedNetwork= transformationManager.getTransformedNetwork(networkId);
 		
-		String nwId = "\""+networkId+"\"";
-		model.addAttribute("networkid",nwId);
-		model.addAttribute("network", network);
+		// Get details of workspace associated with the network using network-workspace mapping details
+		IWorkSpace workspace = network.getNetworkWorkspace().getWorkspace();
+		
+		// Get projectId of the project to which the workspace belongs
+		String projectId = workspaceManager.getProjectIdFromWorkspaceId(workspace.getWorkspaceId());
 		
-		String json = null;
-		if(transformedNetwork!=null){
-			json = d3Creator.getD3JSON(transformedNetwork.getNodes(), transformedNetwork.getLinks());
+		// Get project details using projectId
+		IProject project = projectManager.getProjectDetails(projectId);        
+        
+		// Identify the access type of the project : PUBLIC or PRIVATE
+		EProjectAccessibility projectAccess = project.getProjectAccess();
+		
+		// If the project is PRIVATE, check if the user is allowed to access the information associated with the project
+		// If the project is PUBLIC, allow access to the information associated with the project
+		if(projectAccess == EProjectAccessibility.PRIVATE){
+		    boolean authorizedAccess = false;
+		    
+	        if (projectSecurity.isProjectOwner(userId, projectId)){
+	            authorizedAccess = true;
+	        }
+	        else{
+	            List<String> collaboratorRoles = projectSecurity.getCollaboratorRoles(userId, projectId);
+	            if (collaboratorRoles.contains(RoleNames.ROLE_PROJ_COLLABORATOR_EDITOR) || collaboratorRoles.contains(RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN) || collaboratorRoles.contains(RoleNames.ROLE_PROJ_COLLABORATOR_CONTRIBUTOR)) {
+	                authorizedAccess = true;
+	            }
+	        }
+	            
+	        if(authorizedAccess){
+	            return networkVisualization(networkId, model);
+	        } 
+	        return "public/forbidden";
+		}
+		else {
+		    return networkVisualization(networkId, model);
 		}
-		model.addAttribute("jsonstring",json);
-
-		return "auth/networks/visualize";
 	}
+	
+	private String networkVisualization(String networkId, ModelMap model) throws QuadrigaStorageException{
+	    INetwork network = networkManager.getNetwork(networkId);
+        if(network==null){
+            return "auth/404";
+        }
```

Comment:
this should still be done at the top of ```visualizeNetworks``` (where it was before). If there is not network, the rest of the code does not have to be executed.